### PR TITLE
docs: refactor contributing guidelines - add development setup

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: docs/dev/development-setup.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+- Create the fork on GitHub, using the fork button.
+- Go to the original repo checked out locally (i.e. `$GOPATH/src/github.com/cosmos/ibc-go`)
+- `git remote add fork git@github.com:colin-axner/ibc-go.git`
+
+Now `fork` refers to my fork and `origin` refers to the ibc-go version. So I can `git push -u fork main` to update my fork, and make pull requests to ibc-go from there. Of course, replace `colin-axner` with your git handle.
+
+To pull in updates from the origin repo, run
+
+- `git fetch origin`
+- `git rebase origin/main` (or whatever branch you want)
```
Comment (colin-axner) [inline docs/dev/development-setup.md:18]: maybe we can replace my gh handle with a generic insert `github-handle`? And change `I` to `we`?

File: docs/dev/development-setup.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+- `main` must never fail `make test`.
+- No `--force` onto `main` (except when reverting a broken commit, which should seldom happen).
+- Create a development branch either on `github.com/cosmos/ibc-go`, or your fork (using `git remote add fork`).
+- Before submitting a pull request, begin `git rebase` on top of `main`.
+
+If you open a PR on ibc-go, it is mandatory to update the relevant documentation in `/docs`.
+
+Please make sure to run `make format` before every commit - the easiest way to do this is have your editor run it for you upon saving a file. Additionally please ensure that your code is lint compliant by running `make lint-fix` (requires `golangci-lint`).
+
+All Go tests in ibc-go can be ran by running `make test`.
```
Comment (colin-axner) [inline docs/dev/development-setup.md:74]: I would start with this

File: docs/dev/development-setup.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+```
+{
+    "protoc": {
+        "options": [
+        "--proto_path=${workspaceRoot}/proto",
+        "--proto_path=${workspaceRoot}/third_party/proto"
+        ]
+    }
+}
+```
```
Comment (colin-axner) [inline docs/dev/development-setup.md:59]: I'm not sure if this is true?

**Review by colin-axner — APPROVED** [review]
LGTM, mostly added suggestions on how we can simplify this file. I think it is valuable to consider if there are any parts that can be cut out

File: docs/dev/development-setup.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+We use [Go 1.14 Modules](https://github.com/golang/go/wiki/Modules) to manage dependency versions.
+
+The main branch of every Cosmos repository should just build with `go get`, which means they should be kept up-to-date with their dependencies, so we can get away with telling  people they can just `go get` our software.
+
+Since some dependencies are not under our control, a third party may break our build, in which case we can fall back on `go mod tidy -v`.
+
+## Protobuf
+
+We use [Protocol Buffers](https://developers.google.com/protocol-buffers) along with [gogoproto](https://github.com/gogo/protobuf) to generate code for use in ibc-go.
```
Comment (damiannolan) [inline docs/dev/development-setup.md:30]:
```suggestion
We use [Protocol Buffers](https://developers.google.com/protocol-buffers) along with [buf](https://docs.buf.build/introduction) and [gogoproto](https://github.com/gogo/protobuf) to generate code for use in ibc-go.
```

We will need to change the gogoproto reference for sdk 47 and its using the cosmos/gogoproto fork, right?

File: docs/dev/development-setup.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+For determinstic behavior around protobuf tooling, everything is containerized using Docker. Make sure to have Docker installed on your machine, or head to [Docker's website](https://docs.docker.com/get-docker/) to install it.
+
+For formatting code in `.proto` files, you can run the `make proto-format` command.
+
+For linting and checking breaking changes, we use [buf](https://buf.build/). You can use the commands `make proto-lint` and `make proto-check-breaking` to respectively lint your proto files and check for breaking changes.
```
Comment (damiannolan) [inline docs/dev/development-setup.md:36]: We actually use `buf` now for codegen as well.

File: docs/dev/development-setup.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+To generate the protobuf stubs, you can run `make proto-gen`.
```
Comment (damiannolan) [inline docs/dev/development-setup.md:38]: We could add: to update third party protobuf dependencies, you can run `make proto-update-deps`.

This requires `buf` is installed in the local development environment.

```
brew install bufbuild/buf/buf

# OR

# Substitute PREFIX for your install prefix.
# Substitute VERSION for the current released version.
PREFIX="/usr/local" && \
VERSION="1.9.0" && \
curl -sSL \
    "https://github.com/bufbuild/buf/releases/download/v${VERSION}/buf-$(uname -s)-$(uname -m).tar.gz" | \
    tar -xvzf - -C "${PREFIX}" --strip-components 1
```

File: docs/dev/development-setup.md (hunk 3 above)
Comment (crodriguezvega) [reply docs/dev/development-setup.md:59]: pff, then I will delete it.

File: docs/dev/development-setup.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+Other helpful commands:
+
+- `go get` to add a new go module (including if the existing go module is being semantic version bumped, i.e. my/module/v1 -> my/module/v2).
+- `go get -u` to update an existing dependency.
+- `go mod tidy` to update dependencies in `go.sum`.
+
+## Protobuf
+
+We use [Protocol Buffers](https://developers.google.com/protocol-buffers) along with [gogoproto](https://github.com/cosmos/gogoproto) to generate code for use in ibc-go.
```
Comment (damiannolan) [inline docs/dev/development-setup.md:36]:
```suggestion
We use [Protocol Buffers](https://developers.google.com/protocol-buffers) along with [buf](https://buf.build/) and [gogoproto](https://github.com/cosmos/gogoproto) to generate code for use in ibc-go.
```

**Review by damiannolan — APPROVED** [review]
Approved pending minor changes! 🚀 
Thanks for all the effort on these docs! ❤️

File: docs/dev/development-setup.md (hunk 6 above)
Comment (colin-axner) [reply docs/dev/development-setup.md:44]: Is there buf install documentation we can link to instead?

File: docs/dev/development-setup.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+To pull in updates from the origin repo, run
+
+- `git fetch origin`.
+- `git rebase origin/main` (or whatever branch you want).
+
+## Dependencies
+
+We use [Go 1.14 Modules](https://github.com/golang/go/wiki/Modules) to manage dependency versions.
+
+The main branch of every Cosmos repository should just build with `go get`, which means they should be kept up-to-date with their dependencies, so we can get away with telling  people they can just `go get` our software.
```
Comment (colin-axner) [inline docs/dev/development-setup.md:24]:
```suggestion
The main branch of every Cosmos repository should just build with `go get`, which means they should be kept up-to-date with their dependencies, so we can get away with telling people they can just `go get` our software.
```

Comment (colin-axner) [reply docs/dev/development-setup.md:24]: I'm not sure how useful this section is. I feel like `go get` is something every go module makes use of?

**Review by colin-axner — APPROVED** [review]
ACK, thanks :)

**Review by chatton — COMMENTED** [review]
Looks great, my main concern was the usage of `GOPATH`, please let me know if this is required for some reason.

File: docs/dev/development-setup.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+- Build the `simd` test chain binary with `make build`.
+- `main` must never fail `make test`.
+- No `--force` onto `main` (except when reverting a broken commit, which should seldom happen).
+- Create a development branch either on `github.com/cosmos/ibc-go`, or your fork (using `git remote add fork`).
+- Before submitting a pull request, begin `git rebase` on top of `main`.
+
+All Go tests in ibc-go can be ran by running `make test`.
+
+Please make sure to run `make format` before every commit - the easiest way to do this is have your editor run it for you upon saving a file. Additionally please ensure that your code is lint compliant by running `make lint-fix` (requires `golangci-lint`).
+
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply docs/dev/development-setup.md:66]: Sounds like a good idea! We can investigative this separately. I will leave this for now.

File: docs/dev/development-setup.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+## Forking
+
+Please note that Go requires code to live under absolute paths, which complicates forking. While a fork lives at `https://github.com/<github-handle>/ibc-go`, the code should never exist at `$GOPATH/src/github.com/<github-handle>/ibc-go`. Instead, we use `git remote` to add the fork as a new remote for the original repo `$GOPATH/src/github.com/cosmos/ibc-go`), and do all the work there.
+
+For instance, to create a fork and work on a branch of it, we would:
+
+- Create the fork on GitHub, using the fork button.
+- Go to the original repo checked out locally (i.e. `$GOPATH/src/github.com/cosmos/ibc-go`)
+- `git remote add fork git@github.com:<github-handle>/ibc-go.git`
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply docs/dev/development-setup.md:11]: You're right. I will delete this part.

**Review by chatton — APPROVED** [review]
LGTM!

**Pull request merged**